### PR TITLE
ADMM and FISTA added

### PR DESCRIPTION
**Pull request by straw-boy**
Output of FISTA and ADMM match with SLOPE::SLOPE() if we exclude intercept term. Currently, intial value of \beta is set to be vector with elements 0.1 in FISTA. If we set it to 0 vector, then intercept term matches with SLOPE::SLOPE() for FISTA. Diagnostics not added for ADMM yet.

### CODE REVIEW

**Comment by jolars** [comment]
@straw-boy, could you make @JonasWallin a contributor to the repository as well and assign him to review?

**Comment by straw-boy** [comment]
@jolars I have already sent an invite.

File: src/algorithms/newton_raphson.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    g = gradient(x,y,lin_pred)+rho*(z-u);
+    h = hessian(x,y,lin_pred);
+
+
+    h.diag() += rho;
+
+    mat step = -solve(h,g);
+    double decrement = dot(g,step);
+
+    if( decrement*decrement < tolerance){
```
[the remark this answers is not in the document]
Comment (straw-boy) [reply src/algorithms/newton_raphson.h:41]: You are right that it should be _0.5*decrement < tolerance_ . I just used a rough tolerance value here. I'll set it to exact expression.

**Comment by straw-boy** [comment]
There is a line in src/algorithms/fista.h where I set initial value of beta to something non-zero. Due to this, outputs of solvers::FISTA and SLOPE::SLOPE() don't match for the intercept term when alpha is large. Apart from this, I don't think there is anything else. I'll promptly remove the imported files and functions of SLOPE which are not being used.

**Comment by jolars** [comment]
> There is a line in src/algorithms/fista.h where I set initial value of beta to something non-zero. Due to this, outputs of solvers::FISTA and SLOPE::SLOPE() don't match for the intercept term when alpha is large. Apart from this, I don't think there is anything else. I'll promptly remove the imported files and functions of SLOPE which are not being used.

Hm, the initial value of the coefficients shouldn't make any difference in this regard (unless is wrong).

**Comment by straw-boy** [comment]
I have removed all the source files from SLOPE which I am not using. However, I have kept a lot of R functions that are extremely convenient to compare output with SLOPE::SLOPE().

**Comment by jolars** [comment]
the tests seem to be running very slowly on my computer. Is it possible to speed them up?

**Comment by straw-boy** [comment]
Right now, there is no easy way. Possible short fix for now is maybe skip tests for binomial/possion or instead of full regularization path, test equivalence for some small set of alpha values like c(10,0.1,0)

**Comment by jolars** [comment]
> Right now, there is no easy way. Possible short fix for now is maybe skip tests for binomial/possion or instead of full regularization path, test equivalence for some small set of alpha values like c(10,0.1,0)

Yes, please do so. There is no need to use large data sets in the tests; only make sure you're testing 

1. each family, 
2. standardized and not standardized, and 
3. n > p and p > n.

**Comment by straw-boy** [comment]
Were you able to test that intercept term mismatch?

**Comment by straw-boy** [comment]
> 2. standardized and not standardized,
What did you mean by this?

**Comment by jolars** [comment]
> Were you able to test that intercept term mismatch?

I'm looking into it right now.

> > 1. standardized and not standardized,
> 
> What did you mean by this?

I meant checking consistency when predictors are standardized vs. not standardized. But on second thought, I don't think this is important. In fact, I think that you can just drop all the standardization stuff from this package and assume that the input is standardized since the purpose here is just to benchmark.

**Comment by jolars** [comment]
Another thing: it might be worth it to drop the large data sets from the package. It takes forever to just load the package because of these data sets and it might be easier to just download them for the benchmarks when it is time or keep them in a separate package.

**Comment by straw-boy** [comment]
> But on second thought, I don't think this is important. In fact, I think that you can just drop all the standardization stuff from this package and assume that the input is standardized since the purpose here is just to benchmark.

Is it alright if I leave that as it is? That way, I don't have to worry about separte inputs to SLOPE and solvers' functions. It is easier for me to debug.

**Comment by straw-boy** [comment]
> Another thing: it might be worth it to drop the large data sets from the package. It takes forever to just load the package because of these data sets and it might be easier to just download them for the benchmarks when it is time or keep them in a separate package.

I will use your dataset repository for this then (maybe add custom datasets to it later on as per requirement).

**Comment by jolars** [comment]
Sure (but I don't think you would need to separate anything if you just 
standardize the input before calling your functions).

**Comment by straw-boy** [comment]
I am having trouble understanding why github action checks are failing. I noticed that following lines are missing from log of Run r-lib/actions/setup-r@master :

> installer: Package name is R 4.0.1 for macOS
> installer: Installing at base path /
> installer: The install was successful.

Do you know what is happening?

**Comment by jolars** [comment]
I actually have no idea. Maybe something related to the new version of 
R (4.0.2).

I suggest you either try fiddling with the github actions script or try 
tic(<https://github.com/ropensci/tic>) or usethis::use_travis().

J

**Pull request merged**